diff/fd/hessian.go

```diff
@@ -0,0 +1,186 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package fd
+
+import (
+	"math"
+	"sync"
+
+	"gonum.org/v1/gonum/mat"
+)
+
+// Hessian approximates the Hessian matrix of the multivariate function f
+// at the location x. That is
+//  H_{i,j} = ∂^2 f(x)/∂x_i ∂x_j
+// If dst is not nil, the resulting H will be stored in-place into dst
+// and returned, otherwise a new matrix will be allocated first. Finite difference
+// formula and other options are specified by settings. If settings is nil,
+// the Hessian will be estimated using the Forward formula and a default step size.
+//
+// Hessian panics if the size of dst and x is not equal, or if the derivative
+// order of the formula is not 1.
+func Hessian(dst *mat.SymDense, f func(x []float64) float64, x []float64, settings *Settings) *mat.SymDense {
+	n := len(x)
+	if dst == nil {
+		dst = mat.NewSymDense(n, nil)
+	} else {
+		if n2 := dst.Symmetric(); n2 != n {
+			panic("hessian: dst size mismatch")
+		}
+		for i := 0; i < n; i++ {
+			for j := i; j < n; j++ {
+				dst.SetSym(i, j, 0)
+			}
+		}
+	}
+
+	// Default settings.
+	formula := Forward
+	step := math.Sqrt(formula.Step) // Use the sqrt because taking derivatives of derivatives
+	var originValue float64
+	var originKnown, concurrent bool
+
+	// Use user settings if provided.
+	if settings != nil {
+		if !settings.Formula.isZero() {
+			formula = settings.Formula
+			step = math.Sqrt(formula.Step)
+			checkFormula(formula)
+			if formula.Derivative != 1 {
+				panic(badDerivOrder)
+			}
+		}
+		if settings.Step != 0 {
+			if settings.Step < 0 {
+				panic(negativeStep)
+			}
+			step = settings.Step
+		}
+		originKnown = settings.OriginKnown
+		originValue = settings.OriginValue
+		concurrent = settings.Concurrent
+	}
+
+	evals := n * (n + 1) / 2 * len(formula.Stencil) * len(formula.Stencil)
+	for _, pt := range formula.Stencil {
+		if pt.Loc == 0 {
+			evals -= n * (n + 1) / 2
+			break
+		}
+	}
+
+	nWorkers := computeWorkers(concurrent, evals)
+	if nWorkers == 1 {
+		hessianSerial(dst, f, x, formula.Stencil, step, originKnown, originValue)
+		return dst
+	}
+	hessianConcurrent(dst, nWorkers, evals, f, x, formula.Stencil, step, originKnown, originValue)
+	return dst
+}
+
+func hessianSerial(dst *mat.SymDense, f func(x []float64) float64, x []float64, stencil []Point, step float64, originKnown bool, originValue float64) {
+	n := len(x)
+	xCopy := make([]float64, n)
+	fo := func() float64 {
+		copy(xCopy, x)
+		return f(x)
+	}
+	is2 := 1 / (step * step)
+	origin := getOrigin(originKnown, originValue, fo, stencil)
+	for i := 0; i < n; i++ {
+		for j := i; j < n; j++ {
+			var hess float64
+			for _, pti := range stencil {
+				for _, ptj := range stencil {
+					var v float64
+					if pti.Loc == 0 && ptj.Loc == 0 {
+						v = origin
+					} else {
+						copy(xCopy, x) // perform copy anew because of floating point issues
```

Comment:
Ah, I thought the primary motivation was the protection against modification because otherwise the modified elements could be stored into local variables and we could avoid the allocations and copying. Also, a comment in `Gradient` uses this motivation, so these two should be unified.